### ISSUE TITLE
TGA File has no data when read_image called

### ISSUE THREAD

**Issue by DougRogers** [OpenImageIO/oiio#3277]
```
#include <OpenImageIO/imageio.h>
using namespace OIIO;

int main(int argc, char *argv[])
{
    const char *filename = "F:\\American_cherry.tga";

    auto inp = ImageInput::open(filename);
    if (!inp)
    {
        return -1;
    }
    const ImageSpec &spec = inp->spec();
    int xres              = spec.width;
    int yres              = spec.height;
    int channels          = spec.nchannels;
    std::vector<unsigned char> pixels(xres * yres * channels);
    inp->read_image(TypeDesc::UINT8, &pixels[0]);
    inp->close();

    for (auto p : pixels)
    {
        if (p)
        {
            __debugbreak();
        }
    }
    return 1;
}
```
The breakpoint statment is never reached.

[American_cherry.zip](https://github.com/OpenImageIO/oiio/files/7876270/American_cherry.zip)

**Comment by lgritz** [comment]
Yeah, something odd is up with that file, must hit some edge case. Even `iinfo -v -stats American_cherry.tga` says the whole image is black, which it clearly is not when I display with other programs.

Investigating...

**Comment by lgritz** [comment]
OK, I understand what's happening.

This file:
* Has an alpha channel.
* The alpha value is 0 in all pixels.
* Is a TGA 1.0 file -- only in TGA 2.0 is there an extra header field that explains the meaning of the alpha channel, so this file lacks that.

So without the header field that says that the colors are premultiplied by alpha (which is not the default value, and is described in the Targa spec as if it were an unusual newfangled thing -- so I'm presuming the default interpretation of Targa files should be unassociated alpha / non-premultiplied color), OIIO is assuming that it should do the usual automatic conversion to premultiplied alpha.

If you want to suppress the auto-association and use the raw unpremultiplied colors without regard to the mask, then what you need to do is

    ImageSpec config;
    config["oiio:UnassociatedAlpha"] = 1;
    auto inp = ImageInput::open(filename, &config);

TL;DR: This file, literally interpreted, really is black. It's only if you want to discard the alpha channel and use the colors without regard to the fact that they are masked, that the color channels make any sense.

**Comment by DougRogers** [comment]
Can you detect TGA 1.0 and not do the alpha multiplication?

**Comment by lgritz** [comment]
TGA 1.0 lacks the way to say that they are premultiplied, but I don't think that means we can conclude that the colors are premultiplied.

Here's the table of what the TGA 2.0 field values mean:

    0: no Alpha data included (bits 3-0 of field 5.6 should also be set to zero)
    1: undefined data in the Alpha field, can be ignored 
    2: undefined data in the Alpha field, but should be retained
    3: useful Alpha channel data is present
    4: pre-multiplied Alpha (see description below)
    5 -127: reserved
    128-255: unassigned

    [ lengthy description of choice 4 follows ]

The spec doesn't say explicitly what alpha means in 1.0, but when describing the 2.0 field, the explanation of premultiplied alpha reads as if it's an unusual concept that readers are encountering for the first time. That leads me to suspect that TGA 1.0 files would have generally been assumed to be interpreted as unassociated alpha. Or do you think that the default 1.0 assumption corresponds to 1-2, meaning that the alpha field isn't really alpha and we should ignore it?

**Comment by DougRogers** [comment]
I think the default TGA 1.0 treats alpha as just another data channel and should default to no multiplication.  

The spec for open works, but I read the files this way and adding the spec does not

`
   oiio::ImageSpec configSpec;
    // libRAW configuration
    configSpec.attribute("raw:auto_bright", 0);   // don't want exposure correction
    configSpec.attribute("raw:use_camera_wb", 1); // want white balance correction
#if OIIO_VERSION <= (10000 * 2 + 100 * 0 + 8)     // OIIO_VERSION <= 2.0.8
    // In these previous versions of oiio, there was no Linear option
    configSpec.attribute("raw:ColorSpace", "sRGB"); // want colorspace sRGB
#else
    configSpec.attribute("raw:ColorSpace", "Linear"); // want linear colorspace with sRGB primaries
#endif
    configSpec.attribute("raw:use_camera_matrix", 3); // want to use embeded color profile

    configSpec["oiio:UnassociatedAlpha"] = 1;

    oiio::ImageBuf inBuf(path, 0, 0, NULL, &configSpec);
`

**Comment by lgritz** [comment]
Unassociated means the colors are not premultiplied.

So you think that for TGA 1.0, we should not treat it as unassociated alpha? And therefore *not* premultiply? (That is, treat it as if it was already premultiplied?)

If you have a TGA 1.0 file with raw pixel value (0.5, 0.5, 0.5, 0.5), and you composite it over (0.5, 0.5, 0.5, 1), what do you think the result should be? If that pixel value is premultiplied / associated alpha, the result is (0.75, 0.75, 0.75, 1), but if it's un-premultiplied / unassociated alpha, then the result is (0.5, 0.5, 0.5, 1).

I'll look at the ImageBuf ctr issue.

**Comment by DougRogers** [comment]
<quote>So you think that for TGA 1.0, we should not treat it as unassociated alpha? And therefore not premultiply? (That is, treat it as if it was already premultiplied?
</quote>
Yes.  I think these images were created as RGB and alpha was to be used for something else or not used at all.  

I'm not sure what you mean "and you composite it over (0.5, 0.5, 0.5, 1), "  For this case. if the file has [0.5, 0.5, 0.5, 0.5] I would expect it to load as [0.5, 0.5, 0.5, 0.5].

**Comment by lgritz** [comment]
> Yes. I think these images were created as RGB and alpha was to be used for something else or not used at all.

Interesting. These particular images? Or all TGA 1.0 images that aren't TGA 2.0 with alpha_type == 3?

> I'm not sure what you mean "and you composite it over (0.5, 0.5, 0.5, 1), " For this case. if the file has [0.5, 0.5, 0.5, 0.5] I would expect it to load as [0.5, 0.5, 0.5, 0.5].

If you have one TGA 1.0 file "A.tga" where all pixels are (0.5, 0.5, 0.5, 1), and a second TGA 1.0 file "B.tga" where all pixels are [0.5, 0.5, 0.5, 0.5], what should be the value of "A.tga OVER B.tga" in the Porter/Duff sense of "over"?

It's really easy to mix up the terminology, so I'm asking the question in a "there is only one precise answer" kind of mathematical way to resolve what we think the "meaning" of color and alpha values are in a TGA 1.0 file.

**Comment by lgritz** [comment]
If the answer to [0.5, 0.5, 0.5, 0.5] OVER [0.5, 0.5, 0.5, 1] == [0.5, 0.5, 0.5, 1], that has to mean that you believe that the values in TGA 1.0 files are *unassociated* alpha (i.e., colors are not premultiplied by alpha).

OIIO always converts unassociated alpha in files to associated alpha in memory, unless asked not to. That means that it will multiply the unassociated color channels by the alpha for you, except if you have the configuration hint "oiio:UnassociatedAlpha"  set to 1 to request the original unassociated values. This is the behavior you were seeing.

On the other hand, if you think that in TGA 1.0 files, "alpha" is meaningless, just some other data channel and is not reliably an alpha at all... er, then I guess we should not premultiply the colors when we read them, and the answer to "A over B" is "you can't do compositing operations because there is not a reliable alpha on any TGA 1.0 image." Is that what we think the widespread understanding is/was of TGA files?

**Comment by DougRogers** [comment]
TGA was created and used a long time ago.  It is possible the people just used it as a color format and did not pay attention to alpha.  Anytime there is an image that has all the alpha channels set to zero is suspect.  

I would say in TGA 1.,0 if all the alphas are zero, the alpha is meaningless and should be unassociated.

**Comment by lgritz** [comment]
Yeah, I think I can get behind the heuristic of "TGA 1.0 + all alpha is zero -> assume alpha is meaningless".

Just to be pedantic, though, we should note that for images with associated alpha, alpha==0 is not suspect. It just means that the pixel represents a glow that doesn't block any of the light behind it. Totally normal and useful. And simply not representable by an unassociated alpha image, which is one of the reason that we try so hard to have all internal representations of color+alpha be associated.

**Comment by lgritz** [comment]
The ImageBuf code you quoted above works for me. What version of OIIO are you using?

**Comment by DougRogers** [comment]
I am using version 2.3.10.1.   I will investigate more.  

>Just to be pedantic, though, we should note that for images with associated alpha, alpha==0 is not suspect.
Agreed.

**Comment by lgritz** [comment]
Fix proposed in #3279

**Comment by DougRogers** [comment]
The ImageBuf was caused by an image scaler that was using the alpha values unrelated to OIIO.
